File: day1/part1.cpp
#include <string>
#include <iostream>

int main() {
    // CAPTCHA string
	  std::string captcha = "91211219";
    
    // Sum of all digits that match next digit in the list
    int sum = 0;
    
    // Loop through all digits (except the last) and see if they match the next one
    for (unsigned int i = 0; i < captcha.length() - 1; i++)
    {
        if (captcha[i] == captcha[i+1])
        {
            // Increment the sum by the digit value (converted from char by subtracting ascii value of '0')
            sum += (captcha[i] - '0');
        }
    }
    
    // Compare the first and last digit of the captcha, since it is circular
    if (captcha[0] == captcha[captcha.length()-1])
    {
        sum += (captcha[0] - '0');
    }
    
    // Print the result
    printf ("%d\n", sum);
}

File: day1/part1.rs
fn main() {
    let captcha = "912219";
    
    // Turn the string into an array of bytes, and add indices
    let captcha: Vec<u8> = captcha.bytes().collect();
    let digits = captcha.iter().enumerate();
    
    // Only keep entries that match the adjacent entry
    let digits = digits.filter(|&(i, e)| captcha[(i+1)%captcha.len()] == *e);
    
    // Convert from ascii digit to value of the digit (u32 to avoid overflow final sum)
    let digits = digits.map(|(_, e)| (e-b'0') as u32);
    
    // Get the result
    let digits = digits.collect::<Vec<_>>();
    let sum: u32 = digits.iter().sum();
    println!("{}", sum);
}
